### DISCUSSION TITLE
embed.js: GET https://remark42.host.com/web/embed.js net::ERR_CONNECTION_REFUSED

### DISCUSSION THREAD

**Discussion by OnkelTem** [umputun/remark42#1713] [Q&A] [answered]
## Backend

I run the remark42 server-side binary like this:

```
$ ./remark42.linux-amd64 server --secret=qwerty --url=https://remark42.host.com
remark42 local-20231016T10:52:00
2023/12/24 12:24:56.630 [INFO]  start server on port :8080
2023/12/24 12:24:56.631 [INFO]  root url=https://remark42.host.com
2023/12/24 12:24:56.631 [INFO]  make data store, type=bolt
2023/12/24 12:24:56.631 [INFO]  bolt store for sites [{FileName:./var/remark.db SiteID:remark}], options {Timeout:30s NoGrowSync:false NoFreelistSync:false PreLoadFreelist:false FreelistType: ReadOnly:false MmapFlags:0 InitialMmapSize:0 PageSize:0 NoSync:false OpenFile:<nil> Mlock:false}
2023/12/24 12:24:56.631 [INFO]  make admin store, type=shared
2023/12/24 12:24:56.631 [INFO]  make cache, type=mem
2023/12/24 12:24:56.631 [INFO]  make avatar store, type=fs
2023/12/24 12:24:56.631 [WARN]  no auth providers defined
2023/12/24 12:24:56.631 [INFO]  activate http rest server on :8080
2023/12/24 12:24:56.635 [INFO]  run file server, embedded
2023/12/24 12:24:56.636 [INFO]  start pictures cleanup, staging ttl=7m30s
2023/12/24 12:24:56.636 [INFO]  activate auto-backup for remark under ./var/backup, duration 24h0m0s
```

## Frontend

However, I cannot fetch embed.js:
```
$ curl https://remark42.host.com/web/embed.js
curl: (7) Failed to connect to remark42.host.com port 443 after 140 ms: Connection refused
```

This is how it looks in the browser:
```
GET https://remark42.host.com/web/embed.js net::ERR_CONNECTION_REFUSED
 ```
 
 I reread the docs several times but haven't found any clues.

**Comment by umputun** [comment, marked as answer]
Unless the part of your setup is not provided, I don't see how this is supposed to work. You are running remark42 on port 8080, HTTP, but access it on port 443 HTTPS. Such a configuration may work only in case if you run it behind a reverse proxy of some kind (see [1](https://remark42.com/docs/manuals/nginx/) and [2](https://remark42.com/docs/manuals/reproxy/)) taking care of https termination and request's forwarding 

If you want to run remark42 and expose it directly on https you should set SSL parameters appropriately, i.e. `--ssl.type=auto --ssl.port=443`. In this case you will have remark42 listening on port 443, and handling SSL certificate for you.   _Pls note: this will not work if you already have another web server running on the same box and on the same port_

**Reply by OnkelTem** [reply]
Thanks for the quick reply!

I also realized that my setup looks weird, but I was not sure what exactly I should fix :) Thanks to your response I can now move on.

Currently, this doesn't work:

```
http: TLS handshake error from 176.33.xx.xx:28224: acme/autocert: unable to satisfy "https://acme-v02.api.letsencrypt.org/acme/authz-v3/29722900xxxx" for domain "remark42.host.com": no viable challenge type found
```

I asked chatGPT and it told me that _Let's encrypt_ probably couldn't contact my remark42 host back to verify the token.

The full log is:
```
./remark42.linux-amd64 server --secret=qwerty --url=https://remark42.host.com --ssl.type=auto 
remark42 local-20231016T10:52:00
2023/12/25 10:38:43.348 [INFO]  start server on port :8080
2023/12/25 10:38:43.348 [INFO]  root url=https://remark42.host.com
2023/12/25 10:38:43.348 [INFO]  make data store, type=bolt
2023/12/25 10:38:43.348 [INFO]  bolt store for sites [{FileName:./var/remark.db SiteID:remark}], options {Timeout:30s NoGrowSync:false NoFreelistSync:false PreLoadFreelist:false FreelistType: ReadOnly:false MmapFlags:0 InitialMmapSize:0 PageSize:0 NoSync:false OpenFile:<nil> Mlock:false}
2023/12/25 10:38:43.349 [INFO]  make admin store, type=shared
2023/12/25 10:38:43.349 [INFO]  make cache, type=mem
2023/12/25 10:38:43.349 [INFO]  make avatar store, type=fs
2023/12/25 10:38:43.349 [WARN]  no auth providers defined
2023/12/25 10:38:43.349 [INFO]  activate https server in 'auto' mode on :8443
2023/12/25 10:38:43.353 [INFO]  run file server, embedded
2023/12/25 10:38:43.353 [INFO]  activate http challenge server on port 8080
2023/12/25 10:38:43.353 [INFO]  activate auto-backup for remark under ./var/backup, duration 24h0m0s
2023/12/25 10:38:43.354 [INFO]  start pictures cleanup, staging ttl=7m30s
2023/12/25 10:38:51.792 [WARN]  http: TLS handshake error from 176.33.xx.xx:28224: acme/autocert: unable to satisfy "https://acme-v02.api.letsencrypt.org/acme/authz-v3/29722900xxxx" for domain "remark42.host.com": no viable challenge type found
```

The file "https://acme-v02.api.letsencrypt.org/acme/authz-v3/29722900xxxx" has to following content:
```json
{
  "identifier": {
    "type": "dns",
    "value": "remark42.host.com"
  },
  "status": "deactivated",
  "expires": "2024-01-01T10:38:51Z",
  "challenges": [
    {
      "type": "http-01",
      "status": "pending",
      "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/29722900xxxx/bCJ7lg",
      "token": "Ou_bMd7SCxwLwEdqrZ1Cl0_bbhPZFfcNqhCq2ghHX08"
    },
    {
      "type": "dns-01",
      "status": "pending",
      "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/29722900xxxx/NRTSMQ",
      "token": "Ou_bMd7SCxwLwEdqrZ1Cl0_bbhPZFfcNqhCq2ghHX08"
    },
    {
      "type": "tls-alpn-01",
      "status": "pending",
      "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/29722900xxxx/_w_91g",
      "token": "Ou_bMd7SCxwLwEdqrZ1Cl0_bbhPZFfcNqhCq2ghHX08"
    }
  ]
}
```

Merry Christmas!

**Reply by OnkelTem** [reply]
While I'm waiting for your reply, I decided to try HTTP.  If I run remark42 like this:

```
./remark42.linux-amd64 server --secret=qwerty --url=http://remark42.host.com
```

then I can fetch **embed.js** with CURL. However, it doesn't work in the browser: the request is being blocked and this message appears in the console:

```
publish.js:1 Refused to load the script 'http://remark42.host.com:8080/web/embed.js' because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline' 'unsafe-eval' *.obsidian.md blob: https://*:*". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

I found only one solution to this: adding the appropriate HTTP header or this meta to the HTML:

```
<meta http-equiv="Content-Security-Policy" 
     content="script-src 'self' 'unsafe-inline' 'unsafe-eval' *.obsidian.md blob: https://*:* http://remark42.host.com:8080;">
````

Unfortunately, I cannot do it because my site is an Obsidian Publish one, where I don't have access either to HTTP headers or to HTML.

**Reply by OnkelTem** [reply]
Back to HTTPS.

I talked to chatGPT, and it seems the situation is the following:

1) So I run remark42 like this: `./remark42.linux-amd64 server --secret=qwerty --url=https://remark42.host.com --ssl.type=auto` and it reports that it's listening `:8080` and `:8443`

2) **Let's Encrypt** is probably trying to contact `remark43.host.com` on port 80 to validate the host. Specifically, it might be requesting the URL: `http://yourdomain/.well-known/acme-challenge/<token>` which in my case equals to `http://remark42.host.com/.well-known/acme-challenge/<token>`.

3) However, I have nothing running on port 80 because remark42 is running on `:8080`.

So if the "http-01" validation really happens this way, then I don't understand: shouldn't remark42 also serve port 80 in this case to complete the validation? 

Can the validation be completed manually? 

Btw even in the `--dbg` mode, remark42 doesn’t provide any information about handshaking with "Let's Encrypt."

**Reply by umputun** [reply]
re: http port - see `--port` parameter, which is 8080 by default. Just set it to 80 

Regarding `*.obsidian.md` and custom embdding sources - see `--allowed-hosts` (limit hosts/sources allowed to embed comments)

_pls note: your configuration running remark42 directly (without a docker) and exposing 80/443 to the internet is a little bit unusual. Our defaults for ports and everything else cover common use cases, i.e. remak42 running in a docker container behind the proxy. You may want to check docks and all the documented parameters (running remark42 binary with --h will print them too, for example, `ramark42 server -h`), and this way you can see all the settings you can adjust_

**Reply by OnkelTem** [reply]
Thanks, I'll try it!

I cannot run docker unfortunately because my VPS doesn't support it (too cheap). But I can rent another one if it's needed :)

**Reply by OnkelTem** [reply]
@umputun 

> re: http port - see --port parameter, which is 8080 by default. Just set it to 80

Thanks, it worked! At least the previous error has gone and I could successfully fetch the `embed.js` script with CURL. However, I couldn't fix the issue in the browser with:

```
publish.js:1 Refused to load the script 'http://remark42.host.com:443/web/embed.js' because 
it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline' 'unsafe-eval' 
*.obsidian.md blob: https://*:*". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback.
```

> Regarding *.obsidian.md and custom embdding sources - see --allowed-hosts (limit hosts/sources allowed to embed comments)

I tried setting `--allowed-hosts=*obsidian.md`  and more specific:

```
$ ./remark42.linux-amd64 server --secret=qwerty --url=https://remark42.host.com --ssl.port=443 --ssl.type=auto --port=80 --allowed-hosts='self',host.com,publish.obsidian.md
```

but the error doesn't go away.

**Reply by umputun** [reply]
This is not how you pass multiple hosts, it is a repeatable parameter, i.e. `--allowed-hosts=publish.obsidian.md --allowed-hosts=host.com`

**Reply by OnkelTem** [reply]
The last problem was due to my inattention, but I've already figured that out: I didn't change the url from `http` to `https` on the frontend.

Thank you very much, @umputun!

I believe that documentation can be enhanced, but I don't know how atm. So let me write an article on how I configured comments when I finish the configuration :)

**Reply by paskal** [reply]
Please feel free to submit PRs with documentation updates, you are in the best position to add something useful for them. Happy holidays!
